### ISSUE TITLE
Cannot load dynamic linked library from a statically linked application

### ISSUE THREAD

**Issue by andre2007** [ldc-developers/ldc#3310]
I have a shared object library `lib.d` with this code

```d
import core.stdc.stdio : printf;

extern(C) void dummy()
{
   printf("Hello World!\n");
}
```

and an application `app.d` which loads the shared library at runtime and calls it:

```d
extern(C) nothrow @nogc
{
    alias da_dummy = void function();
}

__gshared {
    da_dummy dummy;
}

import std;
import core.sys.posix.dlfcn : dlclose, dlsym, dlopen, RTLD_NOW;

void main()
{
    string libraryFilePath = buildPath(thisExePath.dirName, "lib.so");
    void* h = dlopen(libraryFilePath.toStringz, RTLD_NOW);
    enforce(h !is null, "Error while loading library");
    dummy = cast(da_dummy) dlsym(h, "dummy");
    dummy();
}
```

I use docker multi stage to compile the application (static) and the shared object.

```dockerfile
FROM dlang2/ldc-ubuntu as base

WORKDIR /dist
COPY lib.d /tmp
COPY app.d /tmp
RUN ldc2 /tmp/lib.d -shared -of /dist/lib.so \
    && ldc2 /tmp/app.d -of /dist/app --static
RUN ["/dist/app"]
```

Docker quiets with 
> The command '/dist/app' returned a non-zero code: 139

It works fine by compiling the application without argument --static.
Is this a bug or is this by intention?

**Comment by kinke** [comment]
This is most likely neither LDC specific (DMD has a `-Xcc` switch now too, so you can validate with `dmd -Xcc=-static ...`) nor a good idea in general - app and .so won't share the same druntime/Phobos (incl. GC) nor the same glibc. I don't know what `-static` entails exactly, but AFAIK, it's not exactly the perfect solution with glibc, and musl would probably be the better choice for static linking, though mixing a statically linked app and a lib depending on shared druntime/Phobos most likely isn't a good idea as mentioned.

**Comment by andre2007** [comment]
In my productive system, only the application is written in D, the SO library is written in C++ I assume. I wrote in this example both in D to have s.th. reproducible. 
I will try also with DMD.

**Comment by andre2007** [comment]
The behavior for dmd is:
- with `-Xcc=-static` : Also fails with error code 139
- withoutht `-Xcc=-static` Also fails with error code 139 but also an additional error message `Aborting from src/rt/sections_elf_shared.d(480) DSO being unregistered isn't current last one`

**Comment by andre2007** [comment]
Dockerfile for dmd

```
FROM dlang2/dmd-ubuntu as base

WORKDIR /dist
COPY lib.d /tmp
COPY app.d /tmp
RUN dmd /tmp/lib.d -shared -of=/dist/lib.so \
    && dmd /tmp/app.d -of=/dist/app

RUN ldd /dist/lib.so | tr -s '[:blank:]' '\n' | grep '^/' | \
   xargs -I % sh -c 'mkdir -p $(dirname ./%); cp % ./%;' \
   && ldd /dist/app | tr -s '[:blank:]' '\n' | grep '^/' | \
   xargs -I % sh -c 'mkdir -p $(dirname ./%); cp % ./%;'

FROM scratch
COPY --chown=0:0 --from=base /dist /
ENV LD_LIBRARY_PATH=/usr/lib:/:/lib:/lib64:/dlang/dmd-2.090.0/linux/lib64:/lib/x86_64-linux-gnu
RUN ["/app"]
```

**Comment by dnadlinger** [comment]
`--static` with LDC will give you a fully statically linked binary (i.e. pass `-static` to the linker). Off the top of my head, I'm not sure whether glibc even supports loading C dynamic libraries when the main executable is statically linked. If the C layer already breaks (memory allocation, …), there is little hope for D applications to work on top of that.

**Comment by andre2007** [comment]
It looks like it is working if I write the SO and the application in pure C:

lib.h
```
#ifndef lib_h__
#define lib_h__
 
extern void dummy(void);
 
#endif  // lib_h__
```

lib.c
```
#include <stdio.h>
 
void dummy(void)
{
    puts("Hello, I am a shared library\n");
}
```

Dockerfile
```
FROM dlang2/ldc-ubuntu as base

WORKDIR /dist
COPY lib.h /tmp
COPY lib.c /tmp
COPY app.c /tmp

RUN gcc -c -Wall -Werror -fpic /tmp/lib.c -o /tmp/lib.o \
  && gcc -shared -o /dist/lib.so /tmp/lib.o

RUN gcc -static /tmp/app.c -ldl -o /dist/app
RUN ["/dist/app"]
```

**Comment by dnadlinger** [comment]
What about threads (especially TLS), memory allocation, etc. though? It's been a while since I've looked at this in detail, so behavior on different OSes blends together a bit in my recollection…

**Comment by dnadlinger** [comment]
(You write above that in your actual use case, the library will be C++. Have you tried loading that from D instead, or a runtime-less D library? Loading a D Shared library into a statically linked D program will not work seamlessly, but that doesn't seem to be what you are after here.)

**Comment by andre2007** [comment]
You are right, my actual use case is calling a shared object library written not in D but in C++.
This causes a non zero error code 139.
